src/coap_tinydtls.c

```diff
@@ -226,124 +230,146 @@ get_psk_info(struct dtls_context_t *dtls_context,
-      memcpy(result, psk_info->identity.s, identity_length);
-      result[identity_length] = '\000';
-
-      coap_session_refresh_psk_key(coap_session, &psk_info->key);
-
-      return identity_length;
     }
-
-    identity_length = 0;
-    /* result_length is max size of the returned identity */
-    psk_len = coap_context->get_client_psk(coap_session, (const uint8_t*)id, id_len, (uint8_t*)result, &identity_length, result_length, psk, sizeof(psk));
-    if (!psk_len) {
-      coap_log(LOG_WARNING, "no PSK identity for given realm or buffer too small\n");
+    else {
+      psk_identity = coap_get_session_client_psk_identity(coap_session);
+    }
+    if (psk_identity == NULL) {
+      coap_log(LOG_WARNING, "no PSK identity given\n");
       fatal_error = DTLS_ALERT_CLOSE_NOTIFY;
       goto error;
     }
-    return (int)identity_length;
+    if (psk_identity->length > result_length) {
+      coap_log(LOG_WARNING,
+               "psk_identity too large, truncated to %zd bytes\n",
+               result_length);
+    }
+    else {
+      /* Reduce to match */
+      result_length = psk_identity->length;
+    }
+    memcpy(result, psk_identity->s, result_length);
+    return result_length;
 #else /* ! COAP_CLIENT_SUPPORT */
     return 0;
 #endif /* ! COAP_CLIENT_SUPPORT */
 
   case DTLS_PSK_KEY:
 #if COAP_CLIENT_SUPPORT
     if (coap_session->type == COAP_SESSION_TYPE_CLIENT) {
-      if (!coap_context || !coap_context->get_client_psk)
-        goto error;
-      identity_length = 0;
-      /* Use psk[] as a scratch area for returning the unused identity */
-      psk_len = coap_context->get_client_psk(coap_session, (const uint8_t*)id, id_len, psk, &identity_length, sizeof(psk), result, result_length);
-      if (!psk_len) {
-        coap_log(LOG_WARNING, "no pre-shared key for given realm or buffer too small\n");
+      psk_key = coap_get_session_client_psk_key(coap_session);
+      if (psk_key == NULL) {
+        coap_log(LOG_WARNING, "no PSK key given\n");
         fatal_error = DTLS_ALERT_CLOSE_NOTIFY;
         goto error;
       }
-      return (int)psk_len;
+      if (psk_key->length > result_length) {
+        coap_log(LOG_WARNING,
+                 "psk_key too large, truncated to %zd bytes\n",
+                 result_length);
+      }
+      else {
+        /* Reduce to match */
+        result_length = psk_key->length;
+      }
+      memcpy(result, psk_key->s, result_length);
+      return result_length;
     }
 #endif /* COAP_CLIENT_SUPPORT */
 #if COAP_SERVER_SUPPORT
-    if (coap_context->get_server_psk) {
-      coap_dtls_spsk_t *setup_sdata;
-      setup_sdata = &coap_session->context->spsk_setup_data;
+    if (coap_session->type != COAP_SESSION_TYPE_CLIENT) {
+      coap_bin_const_t lidentity;
 
-      if (!id)
-        id = (const uint8_t *)"";
+      lidentity.length = id ? id_len : 0;
+      lidentity.s = id ? (const uint8_t*)id : (const uint8_t *)"";
```

Comment:
And here

Reply:
Fixed (and reversed code ordering).

Reply:
Actually, there is no guarentee that `identity `(otherwise, why have `id_len`) is passed in NULL terminated. So, code needs to stand.

Reply:
Correct, sorry.